rfcs/0007-synchronous-commit.md

```diff
@@ -0,0 +1,162 @@
+3. For workloads like logging or stream processing where some data loss is acceptable and performance is paramount, `off` or `local` can be used to optimize throughput.
+
+### RocksDB
+
+Let's look at how RocksDB describes synchronous commit in their documentation:
+
+> #### Non-Sync Mode
+>
+> When WriteOptions.sync = false (the default), WAL writes are not synchronized to disk. Unless the operating system thinks it must flush the data (e.g. too many dirty pages), users don't need to wait for any I/O for write.
+>
+> Users who want to even reduce the CPU of latency introduced by writing to OS page cache, can choose Options.manual_wal_flush = true. With this option, WAL writes are not even flushed to the file system page cache, but kept in RocksDB. Users need to call DB::FlushWAL() to have buffered entries go to the file system.
+>
+> Users can call DB::SyncWAL() to force fsync WAL files. The function will not block writes being executed in other threads.
+>
+> In this mode, the WAL write is not crash safe.
+>
+> #### Sync Mode
+>
+> When WriteOptions.sync = true, the WAL file is fsync'ed before returning to the user.
+>
+> #### Group Commit
+>
+> As most other systems relying on logs, RocksDB supports group commit to improve WAL writing throughput, as well as write amplification. RocksDB's group commit is implemented in a naive way: when different threads are writing to the same DB at the same time, all outstanding writes that qualify to be combined will be combined together and write to WAL once, with one fsync. In this way, more writes can be completed by the same number of I/Os.
+>
+> Writes with different write options might disqualify themselves to be combined. The maximum group size is 1MB. RocksDB won't try to increase batch size by proactive delaying the writes.
+
+Like PostgreSQL, RocksDB provides a `sync` option to control commit behavior and durability guarantees. When `sync = true`, a write is not considered committed until the data is `fsync()`ed to storage.
+
+When `sync = false`, a write is considered committed immediately after the transaction completes, without waiting for the WAL to be written. While the WAL is still buffered in the kernel's page cache, data loss can occur if a crash happens.
+
+Unlike PostgreSQL's `synchronous_commit` which offers multiple levels, RocksDB only provides a simple boolean option. This is because RocksDB is an embedded database and doesn't have the primary/standby architecture that PostgreSQL has.
+
+To optimize synchronous commit performance, RocksDB implements Group Commit, which is a common pattern in WAL-based systems. This mechanism batches multiple writes together into a single, larger WAL write and flush operation, should improve I/O throughput a lot when comparing with multiple small writes.
+
+(SlateDB can implement a similar Group Commit mechanism through its Commit Pipeline, allowing us to batch multiple writes into single WAL operations.)
```

Comment:
Don't we kind of already do this because we batch writes due to `flush.interval` and accrue additional writes in a new WAL `KVTable` while `imm_wal` is being flushed?

Reply:
fixed in bade58a